component/kafka/component.go

```diff
@@ -0,0 +1,511 @@
+	// Mark the consumer as ready
+	close(c.ready)
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (c *consumerHandler) Cleanup(sarama.ConsumerGroupSession) error {
+	// if the strategy is to exit on failure then we cancel the context to exit the whole component
+	// and thus prevent the Consume loop from picking up the message again
+	if c.failStrategy == ExitStrategy {
+		c.cancel()
+	}
+	return nil
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (c *consumerHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for {
+		select {
+		case msg, ok := <-claim.Messages():
+			if ok {
+				log.Debugf("message claimed: value = %s, timestamp = %v, topic = %s", string(msg.Value), msg.Timestamp, msg.Topic)
+				topicPartitionOffsetDiffGaugeSet(c.group, msg.Topic, msg.Partition, claim.HighWaterMarkOffset(), msg.Offset)
+				messageStatusCountInc(messageReceived, c.group, msg.Topic)
+				err := c.insertMessage(session, msg)
+				if err != nil {
+					return err
+				}
+			} else {
+				log.Debug("messages channel closed")
+				return nil
+			}
+		case <-c.ticker.C:
+			c.mu.Lock()
+			err := c.flush(session)
+			if err != nil {
+				return err
+			}
+			c.mu.Unlock()
+		case <-c.ctx.Done():
+			if c.ctx.Err() != context.Canceled {
+				log.Infof("closing consumer: %v", c.ctx.Err())
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) flush(session sarama.ConsumerGroupSession) error {
+	var err error
+
+	if len(c.msgBuf) > 0 {
+		messages := make([]Message, 0, len(c.msgBuf))
+		for _, msg := range c.msgBuf {
+			messageStatusCountInc(messageProcessed, c.group, msg.Topic)
+			ctx, sp := c.getContextWithCorrelation(msg)
+			messages = append(messages, &message{ctx: ctx, sp: sp, msg: msg})
+		}
+
+		for i := 0; i <= c.retries; i++ {
+			btc := &batch{messages: messages}
+			err = c.proc(btc)
+			if err == nil {
+				break
+			}
+
+			if c.ctx.Err() == context.Canceled {
+				return fmt.Errorf("context was cancelled after processing error: %w", err)
+			}
+
+			consumerErrorsInc(c.name)
+			if c.retries > 0 {
+				log.Errorf("failed run, retry %d/%d with %v wait: %v", i, c.retries, c.retryWait, err)
+				time.Sleep(c.retryWait)
+			}
+		}
+
+		if err != nil {
+			for _, m := range messages {
+				trace.SpanError(m.Span())
+				messageStatusCountInc(messageErrored, c.group, m.Message().Topic)
+			}
+			switch c.failStrategy {
+			case ExitStrategy:
+				log.Errorf("exhausted all retries but could not process message(s)")
+				c.err = err
+				return err
+			case SkipStrategy:
+				log.Errorf("exhausted all retries but could not process message(s) so skipping")
+			}
+		}
+
+		for _, m := range messages {
+			session.MarkMessage(m.Message(), "")
+			trace.SpanSuccess(m.Span())
+		}
+
+		if c.commitSync {
+			session.Commit()
+		}
+
+		c.msgBuf = make([]*sarama.ConsumerMessage, 0, c.batchSize)
```

Comment:
We can optimize this part, check out 
https://play.golang.org/p/N8ggnXbxQ2p

Reply:
thanks for the tip! 👍